File: src/main/java/no/utgdev/diy/BeanDefinition.java
package no.utgdev.diy;

import no.utgdev.diy.annotations.Bean;

import java.lang.reflect.Method;

public abstract class BeanDefinition {

    public abstract String getName();
    public abstract Object getObject() throws Exception;


    static class FromMethod extends BeanDefinition {
        private final Method method;

        public FromMethod(Method method) {
            this.method = method;
        }

        @Override
        public String getName() {
            Bean annotation = method.getAnnotation(Bean.class);
            if (annotation != null && annotation.name().length() > 0) {
                return annotation.name();
            }
            return method.getName();
        }

        @Override
        public Object getObject() throws Exception {
            Object obj = method.getDeclaringClass().newInstance();
            return method.invoke(obj);
        }
    }

    static class FromClass extends BeanDefinition {
        private final Class<?> type;

        public FromClass(Class<?> type) {
            this.type = type;
        }

        @Override
        public String getName() {
            return type.getSimpleName();
        }

        @Override
        public Object getObject() throws Exception {
            return type.newInstance();
        }
    }
}

File: src/main/kotlin/no/utgdev/diy/BeanDefinition.kt
package no.utgdev.diy

import no.utgdev.diy.annotations.Bean
import java.lang.reflect.Method

sealed class BeanDefinition {
    abstract val name: String
    abstract val value: Any

    class FromMethod(private val method: Method) : BeanDefinition() {
        override val name: String get() {
            val annotation: Bean? = method.getAnnotation(Bean::class.java)
            return if (annotation != null && annotation.name.isNotBlank()) {
                annotation.name
            } else {
                method.name
            }
        }

        override val value: Any get() {
            val instance = method.declaringClass.getDeclaredConstructor().newInstance()
            return method.invoke(instance)
        }

    }

    class FromClass(private val cls: Class<*>) : BeanDefinition() {
        override val name: String get() = cls.simpleName
        override val value: Any get() = cls.getDeclaredConstructor().newInstance()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is BeanDefinition) return false

        if (name != other.name) return false

        return true
    }

    override fun hashCode() = name.hashCode()
}